Reduce splitstore memory usage during chain walks

vyzo opened the pull request:
Walking the chain was using a `cid.Set`, which can grow quite big (30M+ objects, maybe more) leading to signficant _transient_ memory usage.

This adds an ObjectVisitor interface to use instead of a blanket `cid.Set`. This allows us to use a noop set during compaciton (as we already have the markset to lean on, which will be on-disk with `MarkSetType = "badger"`) and markset-backed ones during warmup/check.

We also reset the walked block set on every iteration, to avoid building up a big one with all the blocks; we can do that because the walk is BFS.

vyzo commented:
Opening as draft until I have tested timing/memory effects in my nodes.

vyzo commented:
warmup is 2x slower, takes about an hour now.

vyzo commented:
marking is somewhat slower also; but on the bright side, there was no watchdog action whatsoever (which was a usual occurence).

vyzo commented:
Marked as ready for review, as node burn was successful.

vyzo commented:
Unified the markset/visitor dichotomy, which opens up interesting possibilities: we can now check+mark atomically using `Visit`.

Follow up: parallelize `walkChain`; yes we can now!

vyzo commented:
rebasd on master.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -46,39 +48,12 @@ func NewBadgerMarkSetEnv(path string) (MarkSetEnv, error) {
 	path := filepath.Join(e.path, name)
 
-	// clean up first
-	err := os.RemoveAll(path)
```
Not related to this patch, but this seems scary...

Stebalien replied:
Maybe fine, but I'd consider giving all "transient" datastores a common "tmp-" prefix.

vyzo replied:
Probably better to just put it in the markSetEnv so that the entire directory is tmp labelled.

vyzo replied:
Also, I extract the open/close to precisly have the `Transient` in the function name to make it clear that this really is temporary data.

vyzo replied:
I added the `".tmp"` suffix to all markset names.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		return err
+	})
+
+	switch err {
```
We could _always_ copy the version, then unlock, then do the rest of this function to simplify the locking a bit.

vyzo replied:
we have to hold the lock while doing the fast checks on the maps.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
 	}
 }
 
+func (s *BadgerMarkSet) Visit(c cid.Cid) (bool, error) {
```
Maybe break this into multiple functions so we can defer unlocks?

vyzo replied:
let me think if it can be somehow be rewritten while maintaining the nice optimistic concurrency properties.

vyzo replied:
I broke it into `tryPending` and `tryDB` methods, reused by `Has` as well.
Still not deferring the unlocks, but it is much simpler now.

vyzo replied:
It doesn't look like it needs to be broken out into multiple functions now imo, as all the logic is in the helper functions and what this method does is lock management and gluing.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		return false, nil
+
+	case badger.ErrKeyNotFound:
+		// need to upgrade the lock to exclusive in order to write; take the version count to see
```
I'd move this case out of the switch statement and use an early return.

vyzo replied:
hid the switch inside the `tryDB` helper method.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		version := s.version
+		s.mx.RUnlock()
+
+		s.mx.Lock()
```
If we move this out of the switch statement, we can defer the unlock.

vyzo replied:
I want to be able to do the write without holding the lock.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+	key := c.Hash()
+	pendKey := string(key)
+
+	checkPending := func() (bool, error) {
```
the behavior here is flipped. I'd expect `checkPending` to return _true_ if a write is pending.

vyzo replied:
yeah, the name is perhaps unfortunate; the return signature is (visit, err) -- perhaps I should just name the return variables to make it clear what it returns.

vyzo replied:
I guess I could rename it to `visitPending` so that it is consistent with its return signature.

vyzo replied:
made it a method called `tryPending` and reversed the signature to return `(has, error)`.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		return visit, err
+	}
+
+	err = s.db.View(func(txn *badger.Txn) error {
+		_, err := txn.Get(key)
+		return err
+	})
```
Introduce a `BadgerMarkSet.has` helper?

vyzo replied:
maybe, let me think a bit about that.

vyzo replied:
I made `tryPending` and `tryDB` helpers.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+	s.seqno++
+	s.writing[seqno] = pend
+	s.pend = make(map[string]struct{})
+	s.writers++
```
Isn't this redundant with `len(s.writing)`?

vyzo replied:
yes it is.

vyzo replied:
removed.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+	pend := s.pend
+	seqno := s.seqno
+	s.seqno++
+	s.writing[seqno] = pend
```
So, with a background worker, we can get rid of this and just have a single "writing" and "pending" set. When the background worker is done flushing the last set, it would check to see if it should flush the next one (e.g., use a channel as a flag).

That _should_ help simplify things a bit.

vyzo replied:
The disadvantage of the background worker is that we can do only one write at the time -- this can do multiple concurrent writes, which is strictly better i think.

vyzo replied:
It also loses the timely error return property.

I guess we could have a worker that spawns goroutines to do writes, and somehow bubble up the error to be checked in the next operation, but this gets more complicated than I'd like it to be.

Stebalien replied:
I _think_ badger will handle write parallelism under the covers.

vyzo replied:
hrm not sure -- I prefer to be explicitly parallel and also keep the ability to return an error at some writer so that we can abort.

On the same time, I don't feel terribly strongly about this and we could use a background worker who spawns writers and bubbles up the error in some way.

Stebalien commented on blockstore/splitstore/splitstore_compact.go:
```diff
@@ -1079,27 +1132,43 @@ func (s *SplitStore) waitForMissingRefs(markSet MarkSet) {
-
-					if mark {
-						return errStopWalk
+					visitor, ok := markSet.(MarkSetVisitor)
```
Can we somehow avoid this cast?

vyzo replied:
Yes we can -- I can widen the `MarkSetVisitor` type to be `ObjectVisitor` + `MarkSet` and then we don't need to cast or bother with the else branch code.

vyzo replied:
done.

Stebalien reviewed (commented):
The general logic looks good, but I feel like we can simplify a bit.

vyzo commented on blockstore/splitstore/markset.go:
```diff
@@ -21,8 +21,15 @@ type MarkSet interface {
+
 type MarkSetEnv interface {
 	Create(name string, sizeHint int64) (MarkSet, error)
+	CreateVisitor(name string, sizeHint int64) (MarkSetVisitor, error)
```
Well, the name is what to call this markset if applicable (for debuging purposes, and also restart cleanup purposes).
The sizeHint is a hint on how many elements to expect; the map markset can pre-allocate the map for example.

I guess I should add some docstring.

vyzo replied:
done.

vyzo commented:
Did a number of simplifications:
- The `MarkSetVisitor` interface was widened to be the sum of `ObjectVisitor` and `MakrSet`, thus obviating the need for a cast in splitstore code.
- Removed redundant `writers` state variable from badger markset.
- Simplified and deduplicated the code for `Has` and `Visit` by introducing `tryPending` and `tryDB` helpers.
- Added `".tmp"` suffix to all transient dbs in the markset directory.

vyzo commented:
pushed forgotten commit.

Stebalien commented:
I don't feel too strongly either. I'll take a look at the new code.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -184,8 +186,53 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+func (s *BadgerMarkSet) put(key string) error {
+	s.pend[key] = struct{}{}
+	if len(s.pend) < badgerMarkSetBatchSize {
+		s.mx.Unlock()
```
:bomb:

Stebalien replied:
Folding this function back into the caller is an acceptable solution. Although I'd consider breaking it up into the part that needs the lock, and the part that doesn't.

vyzo replied:
we have to break it and scope the lock externally; the whole reason for releasing the lock in here is to avoid doing the actual write while holding the lock.

Stebalien replied:
Not holding the lock while writing is good. Releasing a lock across a function boundary is not good.

vyzo replied:
understood, allergic reaction :)
I am changing it to release the lock locally and use a separate write function.

vyzo replied:
done.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -184,8 +186,53 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		defer s.mx.Unlock()
+
+		s.version++
+		delete(s.writing, seqno)
```
Do we really want to do this on failure?

vyzo replied:
should we let them accumulate? It probably doesn't matter at all however, since a failure will quickly bubble up and stop the logic that is using the markset.

Stebalien replied:
It's more correct that way, right?

Stebalien replied:
I mean, we _should_ just end up failing out and not making _any_ changes. But I'd feel _slightly_ safer this way (and it can't hurt).

vyzo replied:
it also means that we can't close and we need a separate count for writers (as it was before).
But fine, I will check the error and not remove it if we fail, and also add a writers count to disambiguate for closing.

vyzo replied:
done.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -184,8 +186,53 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		s.version++
+		delete(s.writing, seqno)
+		if len(s.writing) == 0 {
+			s.cond.Broadcast()
```
waitgroup?

Stebalien replied:
This seems like a case where we're never going to call back into this after the user tries to close, as far as I can tell at least.

Stebalien replied:
Not critical, but I prefer to avoid conditions unless I need them as most people don't know about them (let alone understand how to use them correctly).

vyzo replied:
maybe, but this is cleaner imho -- readers should be educated about condition variables, it's an invaluable tool in concurrency.

Stebalien replied:
Eh, doesn't really matter and it does mean we only track the count in one place.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -104,74 +87,93 @@ func (s *BadgerMarkSet) Mark(c cid.Cid) error {
-		if err := batch.Set([]byte(k), empty); err != nil {
-			return err
+	// we need to upgrade the lock to exclusive in order to write; take the version count to see
+	// if there was another write while we were upgrading
+	version := s.version
+	s.mx.RUnlock()
```
(well, once we avoid dropping the lock in put)

vyzo replied:
the problem with the deferred lock here is that it covers the scope of the actual write in the db, so it is _not_ correct; this is a potentially slow operation that can block other readers/writers.
But ok, I will make it release the lock within the function and use another one for doing the actual write.

Stebalien replied:
Well, the point here was less about the deferred unlock and more about reducing the distance between the lock and the unlock calls. I _completely_ missed the fact that we weren't calling unlock after put the first time through.

Stebalien reviewed (changes requested):
So, _most_ of this patch looks fine. But the locking is still a bit tricker than I'd like. I can live with everything _except_ taking the lock in one function and dropping it in another. That's really the only blocking change.

vyzo commented on blockstore/splitstore/markset.go:
```diff
@@ -21,8 +21,19 @@ type MarkSet interface {
+	// name is a unique name for this markset, mapped to the filesystem in disk-backed environments
+	// sizeHint is a hint about the expected size of the markset
 	Create(name string, sizeHint int64) (MarkSet, error)
+	// CreateVisitor is like Create, but returns a wider interface that supports atomic visits.
```
well, we need this if we want to keep bloom marksets around, as it's pointless to use `Visit` on them.

Stebalien replied:
So, what happens if I specify that I want to use a bloom filter? Does lotus start spamming my log with random errors?

Stebalien replied:
Basically, we need to better distinguish between "something went wrong" and "this isn't supported, move along". That can either be done with multiple interfaces/types, sentinel errors, or a boolean "supported".

vyzo replied:
yeah, it will fail the compaction altogether. Perhaps we should kill the bloom code.

vyzo replied:
Arguably, the best solution here is a `SupportsVisitor` method in the `MarkSetEnv` so that we can fail early (in constructor) if the user specifies the bloom filter for markset.

Stebalien replied:
I'd kill it if we're not using it. We can always revive it from git history.

vyzo replied:
done and done -- have cake, eat it too.

So I added a `SupportVisitor` interface for future extensibility, and kill the unused bloom filter code.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -184,10 +195,70 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
+		return false, 0
+	}
+
+	seqno = s.seqno
+	s.seqno++
+	s.writing[seqno] = s.pend
+	s.pend = make(map[string]struct{})
```
We don't _actually_ have to do this here. Instead, we can return "full" (or even just return the number of pending items). Then, we can call `flush` with no parameters. We can then double check to see if the pending set is over the limit and, if so, flush. Otherwise, someone else wins and we walk away.

vyzo replied:
actually, that's a pretty good idea and it will simplify things quite a bit -- no need for separate `pend` and `writing` sets this way.

vyzo replied:
wait, we can't -- failure semantics. I'll keep it as is as the improvement would be marginal.

Stebalien commented on blockstore/splitstore/markset_badger.go:
```diff
@@ -92,86 +68,123 @@ func (e *BadgerMarkSetEnv) Create(name string, sizeHint int64) (MarkSet, error)
+		has, err = s.tryDB(key)
+		if has || err != nil {
+			s.mx.Unlock()
+			return false, err
```
I'd still consider that code I posted as it should simplify this function a bit (even if you can't defer). But that's not critical.

vyzo replied:
yeah, very small delta; I'll leave it as is.

Stebalien reviewed (approved):
Two non-blocking nits.

vyzo commented:
rebased on master for merge, as it was a bit behind.
